src/sinks/aws_cloudwatch_logs/service.rs

```diff
@@ -0,0 +1,329 @@
+            clients: HashMap::new(),
+            request_settings,
+            client,
+        }
+    }
+}
+
+impl Service<BatchCloudwatchRequest> for CloudwatchLogsPartitionSvc {
+    type Response = CloudwatchResponse;
+    type Error = crate::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: BatchCloudwatchRequest) -> Self::Future {
+        let events_count = req.events.len();
+        let events_byte_size = req.events.iter().map(|x| x.event_byte_size).sum();
+        let key = req.key;
+        let events = req
+            .events
+            .into_iter()
+            .map(|req| InputLogEvent {
+                message: req.message,
+                timestamp: req.timestamp,
+            })
+            .collect();
+        let svc = if let Some(svc) = &mut self.clients.get_mut(&key) {
+            svc.clone()
+        } else {
+            // Concurrency limit is 1 because we need token from previous request.
+            let svc = ServiceBuilder::new()
+                .buffer(1)
+                .concurrency_limit(1)
+                .rate_limit(
+                    self.request_settings.rate_limit_num,
+                    self.request_settings.rate_limit_duration,
+                )
+                .retry(
+                    self.request_settings
+                        .retry_policy(CloudwatchRetryLogic::new()),
+                )
+                .buffer(1)
+                .timeout(self.request_settings.timeout)
+                .service(CloudwatchLogsSvc::new(
+                    &self.config,
+                    &key,
+                    self.client.clone(),
+                ));
+
+            self.clients.insert(key, svc.clone());
+            svc
+        };
+
+        svc.oneshot(events)
+            .map_ok(move |_x| CloudwatchResponse {
+                events_count,
+                events_byte_size,
+            })
+            .map_err(Into::into)
+            .boxed()
+    }
+}
+
+impl CloudwatchLogsSvc {
+    pub fn new(
+        config: &CloudwatchLogsSinkConfig,
+        key: &CloudwatchKey,
+        client: CloudWatchLogsClient,
+    ) -> Self {
+        let group_name = key.group.clone();
+        let stream_name = key.stream.clone();
+
+        let create_missing_group = config.create_missing_group.unwrap_or(true);
+        let create_missing_stream = config.create_missing_stream.unwrap_or(true);
+
+        CloudwatchLogsSvc {
+            client,
+            stream_name,
+            group_name,
+            create_missing_group,
+            create_missing_stream,
+            token: None,
+            token_rx: None,
+        }
+    }
+
+    pub fn process_events(&self, events: Vec<InputLogEvent>) -> Vec<Vec<InputLogEvent>> {
+        let now = Utc::now();
+        // Acceptable range of Event timestamps.
```

Comment:
Is there a reason we can't do this at a higher level? Seems like it'd be more efficient to never hold on to these events in the first place if the filter criteria are hard-coded/known in advance.

Reply:
No reason, I'll move this up so they are filtered out immediately.

Reply:
fixed in ad628ee0140d02cb601f09e1145031ea7bd15895